selfservice/strategy/oidc/strategy.go

```diff
@@ -258,46 +245,74 @@ func (s *Strategy) validateFlow(ctx context.Context, r *http.Request, rid uuid.U
 		return nil, nil, errors.WithStack(herodot.ErrBadRequest.WithReasonf(`Unable to complete OpenID Connect flow because the OpenID Provider did not return the state query parameter.`))
 	}
 
-	var cntnr authCodeContainer
-	if _, err := s.d.ContinuityManager().Continue(r.Context(), w, r, sessionName, continuity.WithPayload(&cntnr)); err != nil {
+	f, err := s.validateFlow(r.Context(), r, x.ParseUUID(stateParam))
+	if err != nil {
 		return nil, nil, err
 	}
 
-	if state != cntnr.State {
-		return nil, &cntnr, errors.WithStack(herodot.ErrBadRequest.WithReasonf(`Unable to complete OpenID Connect flow because the query state parameter does not match the state parameter from the session cookie.`))
+	cntnr := authCodeContainer{}
+	if f.GetType() == flow.TypeBrowser {
+		if _, err := s.d.ContinuityManager().Continue(r.Context(), w, r, sessionName, continuity.WithPayload(&cntnr)); err != nil {
+			return nil, nil, err
+		}
+		if stateParam != cntnr.State {
+			return nil, &cntnr, errors.WithStack(herodot.ErrBadRequest.WithReasonf(`Unable to complete OpenID Connect flow because the query state parameter does not match the state parameter from the session cookie.`))
+		}
+	} else {
+		cntnr.State = stateParam
+		cntnr.FlowID = stateParam
 	}
 
-	req, err := s.validateFlow(r.Context(), r, x.ParseUUID(cntnr.FlowID))
-	if err != nil {
-		return nil, &cntnr, err
+	if errorParam != "" {
+		return f, &cntnr, errors.WithStack(herodot.ErrBadRequest.WithReasonf(`Unable to complete OpenID Connect flow because the OpenID Provider returned error "%s": %s`, r.URL.Query().Get("error"), r.URL.Query().Get("error_description")))
 	}
-
-	if r.URL.Query().Get("error") != "" {
-		return req, &cntnr, errors.WithStack(herodot.ErrBadRequest.WithReasonf(`Unable to complete OpenID Connect flow because the OpenID Provider returned error "%s": %s`, r.URL.Query().Get("error"), r.URL.Query().Get("error_description")))
+	if codeParam == "" {
+		return f, &cntnr, errors.WithStack(herodot.ErrBadRequest.WithReasonf(`Unable to complete OpenID Connect flow because the OpenID Provider did not return the code query parameter.`))
 	}
 
-	if code == "" {
-		return req, &cntnr, errors.WithStack(herodot.ErrBadRequest.WithReasonf(`Unable to complete OpenID Connect flow because the OpenID Provider did not return the code query parameter.`))
-	}
+	return f, &cntnr, nil
+}
 
-	return req, &cntnr, nil
+func registrationOrLoginFlowID(flow any) (uuid.UUID, bool) {
+	switch f := flow.(type) {
+	case *registration.Flow:
+		return f.ID, true
+	case *login.Flow:
+		return f.ID, true
+	default:
+		return uuid.Nil, false
+	}
 }
 
-func (s *Strategy) alreadyAuthenticated(w http.ResponseWriter, r *http.Request, req interface{}) bool {
-	// we assume an error means the user has no session
-	if _, err := s.d.SessionManager().FetchFromRequest(r.Context(), r); err == nil {
-		if _, ok := req.(*settings.Flow); ok {
+func (s *Strategy) alreadyAuthenticated(w http.ResponseWriter, r *http.Request, f interface{}) bool {
```

Comment:
Why are these changes here needed?

Reply:
There are two changes here:

* First, if the user is already authenticated and a code was requested, we want to write the existing session to the code-token exchange, so that the client can fetch the session token in exchange for the code.
* Second, the old behaviour was to always redirect the client to the default return_to URL, not considering the more specific return_to URLs that usually take precedence over the default (return_to in the flow, then flow-specific return_to URL). I treated this as a bug and fixed it for all flows.